### ISSUE TITLE
OIDC expirations cause more frequent expired connections

### ISSUE THREAD

**Issue by kegsay** [matrix-org/sliding-sync#274]
When the poller receives a 401, it calls `OnExpiredToken` which removes the token (not device) from the DB then pokes pubsub with `pubsub.V2ExpiredToken` and the user/device ID affected. This then hits this code:
```
func (h *SyncLiveHandler) OnExpiredToken(p *pubsub.V2ExpiredToken) {
	h.EnsurePoller.OnExpiredToken(p)
	h.ConnMap.CloseConnsForDevice(p.UserID, p.DeviceID)
}
```
Which:
 - Removes the user/device from the ensure polling map, meaning the next time we see a request from this user we need to block on a sync v2 request.
 - Close all SS connections for this user/device.

The latter is troublesome, because it means clients will see M_UNKNOWN_POS on their next request. This then causes the client to do an initial sync again. To provide a seamless transition between tokens, it would be better if the proxy:
 - kept the SS connections open `OnExpiredToken`, but flagged it as expired in some way.
 - when receiving the next client request:
    * if the connect is not expired, continue as normal.
    * if the connection is expired:
        - if the new token is valid (that is, /whoami returns 200 OK) then unexpire the connection and continue, without blocking on ensure polling.
        - else expire the connection and return M_UNKNOWN_TOKEN.

So in other words, instead of eagerly expiring connections when the poller gets a 401, instead lazily expire connections when the next client request gets M_UNKNOWN_TOKEN. In the case of a logout, this next request will never arrive, but the 30min TTL will clean up the connection.

**Comment by S7evinK** [comment]
Given OIDC tokens expire every 5min (iirc), maybe, if we can differentiate between normal Synapse and OIDC tokens, we should decrease the TTL as well. Otherwise we might have 6 expired connections at a time, for one device.
I think normal Synapse tokens start with `syt_`?
